src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs

```diff
@@ -186,22 +186,31 @@ private bool DirectoryExists(string path, out int lastError)
 
         public override IEnumerable<string> EnumeratePaths(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
-            return Win32FileSystemEnumerableFactory.CreateFileNameIterator(fullPath, fullPath, searchPattern,
-                (searchTarget & SearchTarget.Files) == SearchTarget.Files,
-                (searchTarget & SearchTarget.Directories) == SearchTarget.Directories,
-                searchOption);
+            FindEnumerableFactory.NormalizeInputs(ref fullPath, ref searchPattern);
+            switch (searchTarget)
+            {
+                case SearchTarget.Files:
+                    return FindEnumerableFactory.UserFiles(fullPath, searchPattern, searchOption == SearchOption.AllDirectories);
+                case SearchTarget.Directories:
+                    return FindEnumerableFactory.UserDirectories(fullPath, searchPattern, searchOption == SearchOption.AllDirectories);
+                case SearchTarget.Both:
+                    return FindEnumerableFactory.UserEntries(fullPath, searchPattern, searchOption == SearchOption.AllDirectories);
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(searchTarget));
+            }
         }
 
         public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
+            FindEnumerableFactory.NormalizeInputs(ref fullPath, ref searchPattern);
```

Comment:
This would appear to change what exception gets thrown if both searchTarget wasn't a  known enum value and if searchPattern was bad.  Previously we would throw for the searchTarget, no we'll throw for the searchPattern.  We may not care about that ordering, but wanted to call it out, and if no tests failed for it, we might have a tiny test hole to consider patching.

Reply:
I'm not seeing it. I removed asserts, I believe this is still checked at all of the entry points.

Reply:
> I'm not seeing it

Let's assume you had both an invalid searchPattern and an invalid searchExpression.  Previously, you'd enter the method, it would jump to the default in the switch, and it would throw an expression for the searchPattern.  Now, the first thing it does is call NormalizeInputs, which will first check searchExpression, and if it's invalid, it'll throw an exception for searchExpression.  I'm pointing out that we're throwing a different error now with this change.  Whether that matters, eh, up to you.

Reply:
Derp- I was looking at SearchOption, sorry. In this case I think I prefer the order change and don't think it is risky.